library/core/src/iter/adapters/flatten.rs

```diff
@@ -282,6 +314,21 @@ where
         let (flo, fhi) = self.frontiter.as_ref().map_or((0, Some(0)), U::size_hint);
         let (blo, bhi) = self.backiter.as_ref().map_or((0, Some(0)), U::size_hint);
         let lo = flo.saturating_add(blo);
+
+        if let Some(fixed_size) = <<I as Iterator>::Item as ConstSizeIntoIterator>::size() {
+            let (lower, upper) = self.iter.size_hint();
+
+            let lower = lower.saturating_mul(fixed_size).saturating_add(lo);
+            let upper = upper.and_then(|i| i.checked_mul(fixed_size));
+            let upper = fhi
+                .zip_with(bhi, usize::checked_add)
+                .flatten()
+                .zip_with(upper, usize::checked_add)
+                .flatten();
```

Comment:
I'm getting a bit lost in all the combinators here.  Could you rewrite it to just a `try { ... }` block?  They're already used in core for iterators in other places, so shouldn't even need a `feature`.

I think this is something like
```rust
let upper = try {
    fhi?.checked_add(bhi?)?.checked_add(i.checked_mul(fixed_size)?)?
};
```

Reply:
I'll give it a try.